When the WebComponentExporter constructor fails with an exception, Maven will print a non-informative "Could not write exported web component module!" error message

### Description of the bug / feature

The Maven Vaadin plugin needs to create instances of `WebComponentExporter`s in order to obtain the tag name and the default parameter values for the webcomponent/component being exported. This is highly unusual since this causes customer code to be instantiated **at build time** (where the necessary runtime environment is not available), which could cause the customer-specific WebComponentExporter classes to fail with an exception. If this happens, Maven will print a non-informative "Could not write exported web component module!" error message.

We've spent hours trying to figure out what's wrong, and it took inside Vaadin knowledge to learn the origin of the issue.

Since creating customer classes in compile time is unusual, this needs to be clearly documented in WebComponentExporter's javadoc. Also, the error message must be made much more informative.

Excerpt from the stacktrace:
```
Caused by: java.lang.NoClassDefFoundError: java/sql/SQLException
    at be.kava.tarifering.vaadin.views.exporter.QueriesExporter.<init> (QueriesExporter.java:23)
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0 (Native Method)
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance (Constructor.java:490)
    at com.vaadin.flow.internal.ReflectTools.createProxyInstance (ReflectTools.java:483)
    at com.vaadin.flow.internal.ReflectTools.createInstance (ReflectTools.java:451)
    at com.vaadin.flow.component.WebComponentExporterFactory$DefaultWebComponentExporterFactory.create (WebComponentExporterFactory.java:108)
    at com.vaadin.flow.server.webcomponent.WebComponentExporterTagExtractor.apply (WebComponentExporterTagExtractor.java:38)
    at com.vaadin.flow.server.webcomponent.WebComponentModulesWriter.getTag (WebComponentModulesWriter.java:142)
    at com.vaadin.flow.server.webcomponent.WebComponentModulesWriter.writeWebComponentToDirectory (WebComponentModulesWriter.java:114)
    at com.vaadin.flow.server.webcomponent.WebComponentModulesWriter.lambda$writeWebComponentsToDirectory$0 (WebComponentModulesWriter.java:94)
```

### Minimal reproducible example

Create a WebComponentExporter class which throws in the constructor.

### Expected behavior

Maven should fail with an informative error message, such as "Failed to create instances of MyCustomExporter because of NoClassDefFoundError: java/sql/SQLException. Please read WebComponentExporter's javadoc for more information on this. Please run Maven with the -e switch (or Gradle with the --stacktrace switch), to learn the full stack trace.

Also, this needs to be clearly documented in WebComponentExporter's javadoc.

### Actual behavior

"Could not write exported web component module!" is printed. When maven is run with `-e`, the exception stack trace is printed which still doesn't suggest as to why the error happened.

### Versions:
    - Vaadin / Flow version: 14.5.3, 19.0.5

--- a/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java
+++ b/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java
@@ -46,6 +46,7 @@ import com.vaadin.flow.utils.FlowFileUtils;
 import elemental.json.Json;
 import elemental.json.JsonObject;
 import elemental.json.impl.JsonUtil;
+
 import static com.vaadin.flow.server.Constants.CONNECT_APPLICATION_PROPERTIES_TOKEN;
 import static com.vaadin.flow.server.Constants.CONNECT_JAVA_SOURCE_FOLDER_TOKEN;
 import static com.vaadin.flow.server.Constants.CONNECT_OPEN_API_FILE_TOKEN;
@@ -155,7 +156,17 @@ public class BuildFrontendUtil {
             builder.copyResources(adapter.getJarFiles());
         }
 
-        builder.build().execute();
+        try {
+            builder.build().execute();
+        } catch (ExecutionFailedException exception) {
+            throw exception;
+        } catch (Throwable throwable) {
+            throw new ExecutionFailedException(
+                    "Error occured during goal execution: "
+                            + throwable.getMessage()
+                            + "Please run Maven with the -e switch (or Gradle with the --stacktrace switch), to learn the full stack trace.",
+                    throwable);
+        }
 
     }
 
@@ -252,38 +263,51 @@ public class BuildFrontendUtil {
         ClassFinder classFinder = adapter.getClassFinder();
         Lookup lookup = adapter.createLookup(classFinder);
 
-        new NodeTasks.Builder(lookup, adapter.npmFolder(),
-                adapter.generatedFolder(), adapter.frontendDirectory(),
-                adapter.buildFolder())
-                        .runNpmInstall(adapter.runNpmInstall())
-                        .withWebpack(adapter.webpackOutputDirectory(),
-                                adapter.servletResourceOutputDirectory(),
-                                adapter.webpackTemplate(),
-                                adapter.webpackGeneratedTemplate())
-                        .useV14Bootstrap(
-                                adapter.isUseDeprecatedV14Bootstrapping())
-                        .enablePackagesUpdate(true)
-                        .useByteCodeScanner(adapter.optimizeBundle())
-                        .withFlowResourcesFolder(flowResourcesFolder)
-                        .copyResources(jarFiles)
-                        .copyLocalResources(
-                                adapter.frontendResourcesDirectory())
-                        .enableImportsUpdate(true)
-                        .withEmbeddableWebComponents(
-                                adapter.generateEmbeddableWebComponents())
-                        .withTokenFile(BuildFrontendUtil.getTokenFile(adapter))
-                        .enablePnpm(adapter.pnpmEnable())
-                        .withConnectApplicationProperties(
-                                adapter.applicationProperties())
-                        .withConnectJavaSourceFolder(adapter.javaSourceFolder())
-                        .withConnectGeneratedOpenApiJson(
-                                adapter.openApiJsonFile())
-                        .withConnectClientTsApiFolder(
-                                adapter.generatedTsFolder())
-                        .withHomeNodeExecRequired(adapter.requireHomeNodeExec())
-                        .withNodeVersion(adapter.nodeVersion())
-                        .withNodeDownloadRoot(nodeDownloadRootURI).build()
-                        .execute();
+        try {
+            new NodeTasks.Builder(lookup, adapter.npmFolder(),
+                    adapter.generatedFolder(), adapter.frontendDirectory(),
+                    adapter.buildFolder())
+                            .runNpmInstall(adapter.runNpmInstall())
+                            .withWebpack(adapter.webpackOutputDirectory(),
+                                    adapter.servletResourceOutputDirectory(),
+                                    adapter.webpackTemplate(),
+                                    adapter.webpackGeneratedTemplate())
+                            .useV14Bootstrap(
+                                    adapter.isUseDeprecatedV14Bootstrapping())
+                            .enablePackagesUpdate(true)
+                            .useByteCodeScanner(adapter.optimizeBundle())
+                            .withFlowResourcesFolder(flowResourcesFolder)
+                            .copyResources(jarFiles)
+                            .copyLocalResources(
+                                    adapter.frontendResourcesDirectory())
+                            .enableImportsUpdate(true)
+                            .withEmbeddableWebComponents(
+                                    adapter.generateEmbeddableWebComponents())
+                            .withTokenFile(
+                                    BuildFrontendUtil.getTokenFile(adapter))
+                            .enablePnpm(adapter.pnpmEnable())
+                            .withConnectApplicationProperties(
+                                    adapter.applicationProperties())
+                            .withConnectJavaSourceFolder(
+                                    adapter.javaSourceFolder())
+                            .withConnectGeneratedOpenApiJson(
+                                    adapter.openApiJsonFile())
+                            .withConnectClientTsApiFolder(
+                                    adapter.generatedTsFolder())
+                            .withHomeNodeExecRequired(
+                                    adapter.requireHomeNodeExec())
+                            .withNodeVersion(adapter.nodeVersion())
+                            .withNodeDownloadRoot(nodeDownloadRootURI).build()
+                            .execute();
+        } catch (ExecutionFailedException exception) {
+            throw exception;
+        } catch (Throwable throwable) {
+            throw new ExecutionFailedException(
+                    "Error occured during goal execution: "
+                            + throwable.getMessage()
+                            + "Please run Maven with the -e switch (or Gradle with the --stacktrace switch), to learn the full stack trace.",
+                    throwable);
+        }
     }
 
     /**

--- a/flow-server/src/main/java/com/vaadin/flow/component/WebComponentExporterFactory.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/WebComponentExporterFactory.java
@@ -106,10 +106,10 @@ public interface WebComponentExporterFactory<C extends Component>
         public WebComponentExporter<C> create() {
             try {
                 return ReflectTools.createInstance(exporterClass);
-            } catch (IllegalArgumentException e) {
-                if (e.getCause() != null && e.getCause().getClass()
+            } catch (Throwable t) {
+                if (t.getCause() != null && t.getCause().getClass()
                         .equals(InvocationTargetException.class)) {
-                    Throwable cause2 = e.getCause().getCause();
+                    Throwable cause2 = t.getCause().getCause();
                     if (cause2 != null && cause2.getClass()
                             .equals(NullTagException.class)) {
                         throw new IllegalArgumentException(String.format(
@@ -117,11 +117,16 @@ public interface WebComponentExporterFactory<C extends Component>
                                         + "'%s' give null value to "
                                         + "super(String) constructor?",
                                 WebComponentConfiguration.class.getSimpleName(),
-                                exporterClass.getCanonicalName()), e);
+                                exporterClass.getCanonicalName()), t);
                     }
                 }
                 // unknown reason, cannot add information
-                throw e;
+                throw new RuntimeException(
+                        "Couldn't create a new instance of class "
+                                + exporterClass + " because of "
+                                + t.getClass().getSimpleName() + ": "
+                                + t.getMessage(),
+                        t);
             }
         }
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/webcomponent/WebComponentModulesWriter.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/webcomponent/WebComponentModulesWriter.java
@@ -250,10 +250,16 @@ public final class WebComponentModulesWriter implements Serializable {
                         themeName));
                 writeMethod.setAccessible(accessible);
                 return files;
-            } catch (IllegalAccessException | InvocationTargetException
-                    | NullPointerException e) {
+            } catch (IllegalAccessException exception) {
+                throw new RuntimeException("Failed to call '"
+                        + WRITE_MODULES_METHOD
+                        + "' via reflection because Java language access control doesn't allow to call it",
+                        exception);
+            } catch (InvocationTargetException exception) {
                 throw new RuntimeException(
-                        "Could not write exported web component module!", e);
+                        "Could not write exported web component module because of exception: "
+                                + exception.getCause().getMessage(),
+                        exception.getCause());
             }
         }
 

--- a/flow-server/src/test/java/com/vaadin/flow/component/DefaultWebComponentExporterFactoryTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/component/DefaultWebComponentExporterFactoryTest.java
@@ -63,12 +63,15 @@ public class DefaultWebComponentExporterFactoryTest {
 
     @Test
     public void createInnerClass_throws() {
-        exception.expect(IllegalArgumentException.class);
+        exception.expect(RuntimeException.class);
+        exception.expectCause(
+                CoreMatchers.instanceOf(IllegalArgumentException.class));
         exception.expectMessage(
                 CoreMatchers.containsString(InnerClass.class.getName()));
         exception.expectMessage(CoreMatchers.containsString("inner"));
         DefaultWebComponentExporterFactory<Component> factory = new DefaultWebComponentExporterFactory<>(
                 InnerClass.class);
+
         factory.create();
     }
 

